### PR TITLE
WA-250 Create pop-up window to sign transactions

### PR DESCRIPTION
**Pull request by germartinez**
- Create a pop-up to sign the transactions.
- One pop-up for each transaction.
- The pop-up displays the transaction.

TO-DO: Sign transactions.

### CODE REVIEW

File: app/actions/transactions.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+export const ADD_TRANSACTION = 'ADD_TRANSACTION'
```
Comment (apanizo) [inline app/actions/transactions.js]: Question, are those actions directly related to a specific "view" or "route" or are actions used across all the app?

File: app/components/Popup/index.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import React, { Component } from 'react'
+import { connect } from 'react-redux'
+import { Route, withRouter } from 'react-router'
+
+import ConfirmTransaction from 'components/ConfirmTransaction'
+
+class Popup extends Component {
```
Comment (apanizo) [inline app/components/Popup/index.js]: this component could be functional
```
const Popup = () => <ConfirmTransaction />

export default Popup
```

File: app/popup.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import React from 'react'
+import ReactDOM from 'react-dom'
+import { Provider } from 'react-redux'
+import { BrowserRouter } from 'react-router-dom'
+
+import Popup from 'components/Popup'
+import store from 'store'
+
+store
```
Comment (apanizo) [inline app/popup.js]: Why are you creating a new "react app" why not just use react-modal package or something like that in "index.js", what am I missing?

Comment (germartinez) [reply app/popup.js]: The pop-up that is created to confirm transactions is not a modal, but a browser window. "index.js" just represents the chrome extension "window".

File: app/actions/transactions.js (hunk 1 above)
Comment (germartinez) [reply app/actions/transactions.js]: They are used across all the app.

File: app/components/ConfirmTransaction/index.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    this.state = {
+      password: '',
+      errorMessage: '',
+      transaction: {},
+    }
+  }
+
+  componentWillMount = () => {
+    chrome.windows.getCurrent((window) => {
```
Comment (apanizo) [inline app/components/ConfirmTransaction/index.js]: Have you tried to do not place those lines here and move them to the constructor? To me, when the component is going to be mounted, the transaction info is available, so maybe there is no need to use the ```componentWillMount``` hook.

Also if in the constructor and in the ```componentWillMount```hook the state should be created using direct assignment -> ```this.state = ... ```

See:
https://www.andreasreiterer.at/web-development/reactjs-lifecycle-methods
https://engineering.musefind.com/react-lifecycle-methods-how-and-when-to-use-them-2111a1b692b1

File: app/components/ConfirmTransaction/index.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      return false
+    }
+
+    //TO-DO: Confirm transaction
+
+  }
+
+  render() {
+    const { password, errorMessage, transaction } = this.state
+
```
Comment (apanizo) [inline app/components/ConfirmTransaction/index.js]: I would recommend using destructuring ES6 syntax:
```
const { from, to, gas, gasPrice, data } = transaction
```

Not for any reason in particular just for following most common eslint rules. But, this is up to you, of course.

File: app/components/ConfirmTransaction/index.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        <p>Data: {transaction.data}</p>
+
+        <input
+          type='password'
+          placeholder='Password'
+          value={password}
+          onChange={this.updatePassword}
+        />
+
+        {errorMessage &&
```
Comment (apanizo) [inline app/components/ConfirmTransaction/index.js]: Why two lines?

File: app/reducers/transactions.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      newState = [...state]
+      newState.push({
+        tx: action.tx,
+        popupId: action.popupId,
+      })
+      return newState
+
+    case REMOVE_TRANSACTION:
+      newState = [...state]
+      return newState.filter(t => t.popupId !== action.popupId)
```
Comment (apanizo) [inline app/reducers/transactions.js]: filter returns a new copy of the array: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

So ```state.filter(....)``` should work out of the box and code has better performance.

File: app/components/ConfirmTransaction/index.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  showTransaction = (windowId) => {
+    const { transactions } = this.props
+
+    if (transactions && transactions.length > 0) {
+      const ts = transactions.filter(t => t.popupId === windowId)
+
+      if (ts.length === 1 && ts[0].tx) {
+        const tx = ts[0].tx
+
+        this.setState({
```
Comment (apanizo) [inline app/components/ConfirmTransaction/index.js]: Calling setState from the constructor is not allowed, why not just showTransaction returns the transaction and you specify the ```this.state = ... ``` inside... ```getCurrent``` function?

Comment (germartinez) [reply app/components/ConfirmTransaction/index.js]: `getCurrent` function is asyncronous so I'm moving it to `componentDidMount`

Comment (apanizo) [reply app/components/ConfirmTransaction/index.js]: That makes sense.

File: app/components/ConfirmTransaction/index.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+  componentDidMount = () => {
+    chrome.windows.getCurrent((window) => {
+      this.showTransaction(window.id)
+    })
+  }
+
+  showTransaction = (windowId) => {
+    const { transactions } = this.props
+
+    if (transactions && transactions.length > 0) {
```
Comment (apanizo) [inline app/components/ConfirmTransaction/index.js]: I think it is clearer if you isolate the code for checking conditions. Something like:
```
if (!transactions || transactions.length === 0) {
  return
}
```

In that way the core code is not spaced and easier to read and to understand.

File: app/components/ConfirmTransaction/index.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+      const ts = transactions.filter(t => t.popupId === windowId)
+
+      if (ts.length === 1 && ts[0].tx) {
```
Comment (apanizo) [inline app/components/ConfirmTransaction/index.js]: You can rename here variables to make the code easier to read. Something like:
```
const windowTransaction = transactions.filter(...)
const hasTransaction = windowTransaction.length === 1 && windowTransaction[0].tx

if (!hasTransaction) {
  return
}

.....
```

**Pull request merged**